src/session_state.rs

```diff
@@ -4,27 +4,33 @@
 use embedded_nal::IpAddr;
 use heapless::{String, Vec};
 
-pub struct SessionState {
+use embedded_time::{duration::Extensions, Clock, Instant};
+
+pub struct SessionState<C: Clock> {
     // Indicates that we are connected to a broker.
     pub connected: bool,
-    pub keep_alive_interval: u16,
+    pub keep_alive_interval: Option<u16>,
```

Comment:
Let's take `Option<Seconds>` here and have `set_keep_alive_interval` also take `Into<Seconds>` or similar to provide a nice API. That then also cleans up the conversion and scaling in `ping_is_due`.